db/range_del_aggregator.cc

```diff
@@ -354,6 +340,16 @@ Status RangeDelAggregator::AddTombstone(RangeTombstone tombstone) {
         }
       }
 
+      // If we're looking at either the last new tombstone or the last existing
```

Comment:
The previous code did `new_to_old_end_cmp = 1` which would only advance the tombstone map iter, right?

I'd probably just have to sit down and *really* read this code, but I don't understand why the condition here is symmetric and always results in the *new* tombstone ending before the *existing*. If `new_range_dels_iter_end` is not null, but `tombstone_map_iter_end` is `nullptr`, isn't the situation that the new tombstone "sticks out" of the existing one?

Reply:
> The previous code did new_to_old_end_cmp = 1 which would only advance the tombstone map iter, right?

Correct. That resulted in looking in every existing tombstone after the tombstone getting added.

> I'd probably just have to sit down and really read this code, but I don't understand why the condition here is symmetric and always results in the new tombstone ending before the existing. If new_range_dels_iter_end is not null, but tombstone_map_iter_end is nullptr, isn't the situation that the new tombstone "sticks out" of the existing one?

Yeah, I went back and forth on the clearest way to express this. The way to think about it—I think—is that the `new_range_dels_iter_end == nullptr` means the new tombstone is just one point wide. Suppose we're adding tombstone [a, f) when tombstone [e, j) already exists:

```
@2  a----f
@1     e----j
```

The existing tombstones are [e, j) @ 1 and [j, nullptr) @ 0. The new tombstones are [a, f) @ 2 and [f, nullptr) @ 0. As before, [j, nullptr) @ 0 is infinitely wide. I'm claiming that, instead of also pretending that [f, nullptr) @ 0 is infinitely wide, we can pretend it's exactly one point wide, i.e., [f, f] @ 0, because it serves only to mark the endpoint of the _real_ tombstone [a, f) @ 2.

Anyway, as soon as we see a single-point (new) tombstone, we _know_ its end key is less than whatever existing tombstone we're comparing it against. Why? Well we're guaranteed that a) any two tombstones that the loop compares are overlapping, and b) existing tombstones are not single points. There's no way to satisfy those guarantees unless the existing tombstone's upper bound is greater than the single-point (new) tombstone.

Here are the four cases spelled out:

| new_range_dels_iter_end | tombstone_map_iter_end | explanation |
|-|-|-|
| not null | not null | This case is easy: compare the two ends directly. |
| not null | null | Also easy: a null existing tombstone end means the tombstone extends infinitely to the right, so it's obviously ends after any finite range deletion tombstone. |
| null | not null | New tombstone end is less by the logic above. |
| null | null | New tombstone end is less by the logic above. |

This is all very confusing because this method is written as if it's going to merge multiple points at once. @ajkr back in https://github.com/facebook/rocksdb/pull/1614#discussion_r91577557 you wrote:

> one of the motivations for using a vector of new points is that, in a future optimization, we're considering merging multiple new points simultaneously (e.g., all points from a snapshot stripe in an SST file that we just read), which should be faster than merging new points one-at-a-time.

Is that still your eventual plan?

Reply:
Good find. We should eventually make use of the ordering of range deletions in the meta-block (within each snapshot stripe). But doing this is not on any foreseeable roadmap, and I hope it'll be made unnecessary by Cockroach's optimizations :). So feel free to get rid of this overly general code if you wish.